### DISCUSSION TITLE
Fast full-text search or LIKE query on Parquet file on S3

### DISCUSSION THREAD

**Discussion by collimarco** [apache/datafusion#6545] [General]
I am trying to filter some logs that are stored in Parquet format on S3.

I use a query like this:

```
SELECT status, message FROM sample WHERE status = 200 AND message LIKE '%foo%'
```

**Is this the right approach to search some keywords in a text field? Or is there a better/faster function to use instead of SQL LIKE?**

In particular I would like to optimize the speed of the query and reduce the data downloaded (e.g. leveraging the condition pushdown).

**Comment by tustvold** [comment]
You could try enabling filter pushdown (late materialization) for parquet - https://docs.rs/datafusion/latest/datafusion/config/struct.ParquetOptions.html

**Reply by collimarco** [reply]
Thanks! How can I set them? Currently I use `ctx.register_parquet("sample", &path, ParquetReadOptions::default()).await?;` but this code expects only `ParquetReadOptions` and not `ParquetOptions`. Where can I set `ParquetOptions`? is there an example?

**Reply by tustvold** [reply]
You can provide them in the constructor for SessionContext - https://docs.rs/datafusion/latest/datafusion/execution/context/struct.SessionContext.html#method.with_config

**Reply by alamb** [reply]
If this works for you @collimarco  it would be awesome if you could give us some feedback -- we have been contemplating enabling this feature by default for a while (https://github.com/apache/arrow-datafusion/issues/3463) as a way to have better out of the box performnace

**Reply by collimarco** [reply]
@tustvold Is there any code snippet / example? I am new to Rust and I cannot find the right code... already tried several solutions but none works. For example you suggest `SessionContext::with_config(options)` but that raises `expected SessionConfig, found ParquetOptions` when I pass the `ParquetOptions`. @alamb Yes, I would be happy to share the benchmark results on my data if I make it to work. In any case I think it would be useful to enable predicate pushdown by default, in particular for files stored on S3 (to reduce download size).

**Reply by alamb** [reply]
Here is an example for creating a `SessionConfig`: https://github.com/influxdata/influxdb_iox/blob/f571aeb4457b6436fc1e82b7292aa5f54101abe0/datafusion_util/src/config.rs#L18-L30

Then you create a SessionContext like

```rust
// create config as above
let config = ...;

/// create a session context from that config
let ctx = SessionContext::with_config(config);
```

**Reply by collimarco** [reply]
@alamb Thanks it works! Here's the full code for future reference:

```rust
let mut options = ConfigOptions::new();
options.execution.parquet.pushdown_filters = true;
options.execution.parquet.reorder_filters = true;
let config = SessionConfig::from(options);
let ctx = SessionContext::with_config(config);
```

Unfortunately the performance on a Parquet file of 100MB stored on Cloudflare R2 (it's like S3) doesn't change much enabling the condition pushdown. I have made some filter attempts and both versions take 2.5-3 seconds. Maybe the distribution of values in my Parquet file doesn't allow to discard many row groups. Or the `LIKE '%foo%'`query makes it impossible to use the metadata to discard the row groups. 

Is there any way to see some statistics (e.g. total downloaded data, number of rows discarded by pushdown conditions, etc.)?

**Reply by alamb** [reply]
> Is there any way to see some statistics (e.g. total downloaded data, number of rows discarded by pushdown conditions, etc.)?

You can run your query using `EXPLAIN ANALYZE SELECT ... ` which will print out various execution metrics that should tell us how much time is spent in various parts of the execution as well as how many rows are read / pruned and how long it takes to fetch the time

**Reply by collimarco** [reply]
Here's the results.

Normal:

```
Plan with Metrics | CoalesceBatchesExec: target_batch_size=8192, metrics=[output_rows=1, elapsed_compute=427.275µs]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
FilterExec: status@0 = 200 AND message@1 LIKE %foo%, metrics=[output_rows=1, elapsed_compute=118.459867ms]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |
ParquetExec: file_groups={12 groups: [[sample_zstd.parquet:0..2311795], [sample_zstd.parquet:2311795..4623590], [sample_zstd.parquet:4623590..6935385], [sample_zstd.parquet:6935385..9247180], [sample_zstd.parquet:9247180..11558975], ...]}, projection=[status, message], predicate=status@5 = 200 AND message@12 LIKE %foo%, pruning_predicate=status_min@0 <= 200 AND 200 <= status_max@1, metrics=[output_rows=100000, elapsed_compute=12ns, file_open_errors=0, row_groups_pruned=0, predicate_evaluation_errors=0, file_scan_errors=0, pushdown_rows_filtered=0, bytes_scanned=24592966, num_predicate_creation_errors=0, page_index_rows_filtered=0, time_elapsed_opening=5.095043771s, pushdown_eval_time=24ns, time_elapsed_scanning_total=943.016365ms, time_elapsed_processing=164.28534ms, time_elapsed_scanning_until_data=753.595527ms, page_index_eval_time=1.315µs]
```

With condition pushdown:

```
Plan with Metrics | CoalesceBatchesExec: target_batch_size=8192, metrics=[output_rows=1, elapsed_compute=443.608µs]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |
FilterExec: status@0 = 200 AND message@1 LIKE %foo%, metrics=[output_rows=1, elapsed_compute=72.509µs]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
ParquetExec: file_groups={12 groups: [[sample_zstd.parquet:0..2311795], [sample_zstd.parquet:2311795..4623590], [sample_zstd.parquet:4623590..6935385], [sample_zstd.parquet:6935385..9247180], [sample_zstd.parquet:9247180..11558975], ...]}, projection=[status, message], predicate=status@5 = 200 AND message@12 LIKE %foo%, pruning_predicate=status_min@0 <= 200 AND 200 <= status_max@1, metrics=[output_rows=1, elapsed_compute=12ns, num_predicate_creation_errors=0, bytes_scanned=24592966, pushdown_rows_filtered=99999, page_index_rows_filtered=0, predicate_evaluation_errors=0, row_groups_pruned=0, file_scan_errors=0, file_open_errors=0, time_elapsed_processing=178.523739ms, time_elapsed_scanning_total=850.179702ms, pushdown_eval_time=19.6539ms, time_elapsed_opening=5.674377264s, time_elapsed_scanning_until_data=849.057695ms, page_index_eval_time=688ns]
```

**Reply by collimarco** [reply]
Actually you can see that there is a `pushdown_rows_filtered ` greater than zero when condition pushdown is enabled. However other values like `row_groups_pruned ` are still zero (no row group discarded from download?).

**Reply by tustvold** [reply]
> other values like row_groups_pruned are still zero

The row group only stores the min and max values for each column. This means only the status predicate could be used for row group pruning, and for your data it would appear this doesn't eliminate any row groups from consideration

> time_elapsed_scanning_until_data=849.057695ms

This would somewhat suggest that the latency to R2 is pretty high. You might be able to experiment with setting a metadata_size_hint to reduce the number of roundtrips, but in general if this is how long it takes to communicate with R2, this is going to limit what query latency is possible.

> page_index_eval_time=688ns

Has the parquet file been written with a page index? Perhaps you might run it through https://github.com/apache/arrow-rs/blob/master/parquet/src/bin/parquet-layout.rs? This would significantly reduce the ability to do IO pushdown or any form of predicate pruning.

**Reply by collimarco** [reply]
> The row group only stores the min and max values for each column.

I know that it wouldn't be easy to improve... but what about keeping also a Bloom filter for each term in the column (and not just min/max)?

I see that Parquet already supports Bloom filters.

For my understanding, it currently uses them only on the entire value.

Fo example, if I have a column "MovieTitle":

- "The title of my movie"
- "Another movie title"
- "The best movie title"
- ...

Then the current Bloom filters can be used to find only the column chunks/pages that match an exact title. For example you can use the bloom filter to search for "The best movie title".

It would be interesting to have *a bloom filter on the specific words*, instead of using the entire value: in this way you can search the word "best" in the "MovieTitle" column and find the titles that contain that specific word in an efficient way.

It would enable a sort of full-text search of keywords inside text columns. It would also allow predicate pushdown for searches based on keywords.

Would make sense to have such an addition? Where can you make a proposal?

**Reply by collimarco** [reply]
> Has the parquet file been written with a page index?

I don't know. Can you explain?

I created it with the [official Arrow library for Ruby](https://github.com/apache/arrow/tree/main/ruby):

```ruby
arrow_table = Arrow::Table.new(table)
arrow_table.save('sample.parquet')
```

**Reply by tustvold** [reply]
I suspect supporting native indexing for accelerated full text search is likely out of scope for the parquet project, especially since it would be tricky to get consensus on an approach and then drive ecosystem adoption, but you could try proposing the idea on the parquet mailing list. I suspect the suggestion would be to maintain some form of secondary-index like https://github.com/valeriansaliou/sonic or Lucene on top of your parquet files.

Taking a step back though, the bottleneck in your queries appears to be IO to R2, something that additional indexing is unlikely to help dramatically with. I'm not very familiar with R2, but typically object stores have very high first byte latencies after which throughput is very high - consequently the relationship between bytes scanned and latency is often extremely sub-linear.

**Reply by collimarco** [reply]
>  what about keeping also a Bloom filter for each term in the column (and not just min/max)?

Maybe I could convert the text column in an array of words. Example: `"The best movie title"` becomes `["The", "best", "movie", "title"]`

If I make this transformation and I search for a specific term, can Parquet / Datafusion use a condition pushdown? Do array columns have some advantages for term search? Or they need to be scanned entirely like normal text?

**Reply by alamb** [reply]
> If I make this transformation and I search for a specific term, can Parquet / Datafusion use a condition pushdown? Do array columns have some advantages for term search? Or they need to be scanned entirely like normal text?

I don't think DataFusion would be able to help here (as it only has min/max values per column, so even if you split a column into arrays the index would still only have two (min/max) array values  on each row group and data page.

As @tustvold mentions above, supporting fast full text search will likely need a specialized index built on top of your parquet files (I think the academic term is "inverted index") but parquet doesn't support such indexes natively
